src/metabase/api/session.clj

```diff
@@ -97,16 +97,46 @@
          {:status-code 400
           :errors      {:password password-fail-snippet}}))))
 
+(defn- source-address
+  "The `public-settings/source-address-header` header's value, or the `(:remote-addr request)` if not set."
+  [request]
+  (if-let [source-address-header (public-settings/source-address-header)]
+    (if-let [header-value ((:headers request) source-address-header)]
+      header-value
+      (:remote-addr request))
+    (:remote-addr request)))
```

Comment:
You can make this a lot more concise with a bit of destructuring and `or`, e.g.

```clj
(defn- source-address
  [{:keys [headers remote-addr]}]
  (or (some->> (public-settings/source-address-header) (get headers))
      remote-addr))
```

You could even do

```clj
(defn- source-address
  [{{source-addr (public-settings/source-address-header)} :headers, :keys [remote-addr]}]
  (or source-addr remote-addr))
```

but if `source-address-header` is nil and for some insane reason there's a nil value in headers (e.g. `{nil "blah blah blah"}`) you'll get that value back as `source-addr`, so I would go with the former